### ISSUE TITLE
[Bug] brushEnd event yields incorrect coordRange when using dataZoom

### ISSUE THREAD

**Issue by michaelhoarau** [apache/echarts#19117]
### Version

5.4.3

### Link to Minimal Reproduction

https://codepen.io/michoara/pen/poqwQpQ

### Steps to Reproduce

1. Open the CodePen above, show the console
2. Enable multiple brushes and paint a few brushes (at least one)
3. The consoles shows that the current start and end dates associated to the painted brushes. See screenshot from CodePen below:

![screenshot_1](https://github.com/apache/echarts/assets/12375366/3d3f7f5d-ce18-4b93-a094-c1dede4717c9)

5. Use the datazoom to bring the painted brushes out of the current view
6. Paint a new brush (while keeping the others, do not clear them)
7. The consoles shows that the last brush as correct coordinates, while all the others have a start and end dates squashed to 0s with a date/time corresponding to either the start or the end of the datazoom depending on how you zoomed. See screenshot below:

![screenshot_2](https://github.com/apache/echarts/assets/12375366/2f9c2acd-f58e-4a09-8081-0d0e02ddfa7e)

### Current Behavior

When using the datazoom some brushes may slip out of the current view: when that happens, the coordRange for these brushes are not retained and become rather random.

### Expected Behavior

Keep a consistent coordRange for each brush defined, whichever position or zoom the user uses.

### Environment

```markdown
- OS: Windows 10 21H2 Build 9044.3324
- Browser: Chrome 116.0.5845.188 or Firefox 115.2.0esr
```

### Any additional comments?

Users may need to zoom in and out to paint precise brushes. The only workaround I have for now is to remove the datazoom and let the user draw the brushes but this is not practical as I'm dealing with time series data than can span 1-2 years at 1 minute sampling rate.

**Comment by helgasoft** [comment]
Bug is valid, here is a simple workaround. 
Each _brushEnd_ event provides **one** new area (at the end of the list). 
Instead of relying on ECharts, we could keep track by saving it inside our own array.
Result - allAreas with correct ranges.
```
allAreas = [];
chart.on('brushEnd', function(params) {
   newArea = params['areas'].slice(-1)[0];  // get last one only
   allAreas.push[newArea];
}
```

**Comment by michaelhoarau** [comment]
Thanks, this is the workaround I already set in place.
Unfortunately, this won't be enough, as I also have to make the brushes non-transformable, which makes the user experience quite annoying in my app, to say the least. When all the areas have already been squashed, it's very hard to pinpoint which one was was modified. Is there anyway to some sort of unique immutable ID for the different brushes when the brushEnd event is fired?

**Comment by helgasoft** [comment]
> have to make the brushes non-transformable...
When all the areas have already been squashed, it's very hard to pinpoint which one was modified. 

sorry but cannot quite follow...

**Comment by michaelhoarau** [comment]
Let me clarify with an animation:

![echart-brush-issue](https://github.com/apache/echarts/assets/12375366/79929964-046d-4cc3-9221-3f1896214ae1)

In this capture:
* I define 3 brushes
* Use datazoom to move to another part where the 3 initial brushes are not visible
* I then define a 4th brush: this is where the start/end range of the initial 3 brushes become random and are not usable anymore.

Until this step, I can use the workaround you mention

* After that, I use the dataZoom to bring back one of the previous brush
* I then modify it

This is where I get stuck: now, it's not only the "last" brush that gets modified, but another one. I would need to compare the list of brushes immediately before and after this modification: I have a brushEnd event, but not a brushStart. And processing all the brush events is going to be quite overkill.

**Comment by helgasoft** [comment]
good clear explanation, thanks.
Found out that of out-of-zoom areas have identical _coordRange_ values. This simplifies the task of identifying which area has been **updated** - [Demo Code](https://echarts.apache.org/examples/en/editor.html?c=line-simple&code=MYewdgzgLgBFAMMC8MAiBDKBTAdGEA7gBQCUA3ALABQoksAJpusjAIIBO76AnjgGbsQAWyIBvADZYwAcygALAFwwArPAC-AGhhEskoVKhaAlmHpYAHiWQA-GAG0EMANQwAbPABUARni-PJs3MtAFlMOX5xEBB2IlD5HC5TYVIYDxgfeBIAXXJqEAAHKCNwJFFqGBhzVnMjCCVROG58rCUAciL9VphNcpgAI3YAVwhFGDKqCoqoKPE-kHMlO1bxEywADVatVoBrLCx8zZhW4El0dlasjV6KqpqIAElTCyV4K4nJgeG5YJAzNqFBuIivlJJtrnA5IIoFBJAAVJotI5mOaDMDALBg949d6MKDoABaUSE9WxFW41Vq9UazTaADd0OJBhjum8KhAsOwjFg6vYGrj0Ep-VooAi2iswMy1FlqGpqNR6ewYAzxBwsOgICw7FkyBUAPS6pXsNUa3aFGDAQacAzUITcADCcjOUBw4CIrU-IwAoqZDnxUcAiq78lZxhUjHxtPklmdjRccJIZPIYLZlar1fGpLI5CHJjB9Ur6PRwRKCGmWFHWjH1XGICt0UQALReEh2eBZHCgaL0ABK6BkWB1eYN0iwsHE6tg4HE3HBqaN6fyXyIJbTuSxMF6unZY0m-aEv3DM_eFQrVYgcb40U96GAciIRBAGiMViQtlDuaHMD4RiBHJgtNqIw-kkGApyPD8YHDbQQA7KJ2B7PsR1bLIYAAQhQGDO3g3t-zsLwcjGcFcygh9YK7HCRxwaYAGUoE5GQUnQpVxBVecIDsIx2xouiTGkUgQyIj9aAgEBJHjEA-NaQZ8lxLB6BgABiTYiCMJxmycVoFAAHXwcR6CQVonDnY0OKyDSdJLAynEwuCEP7KxBNzYz1VMlgbPIxCB0c7pBNlY9uhIXo_LUcggA).
Next step is to find the event when all areas are **deleted** and clear _allAreas_. Please share if you pull it off.

**Comment by michaelhoarau** [comment]
Hi,

Thanks for your help: I tried to do that, but it's not capturing the right values unfortunately.

When you update a brush while other brushes are not visible, all the non-visible brushes get identical coordRange values, that's correct (well, not intended, but that's a correct statement).
When you use the dataZoom to make them visible again, they keep these identical coordRange, so if you change one of them, you have no way of knowing one was actually changed (as both some out-of-zoom and in-zoom areas now have identical coordRange).

In addition, sometime the coordRanges are not fixed anymore: on the following videos you will see the following:

![echart-brush-end-issue](https://github.com/apache/echarts/assets/12375366/62a50c6f-a790-4dd0-951f-94ba0a820817)

1. I select 4 areas
2. I use the zoom to bring the first 3 out of zoom
3. I update the 4th area width
4. I then reuse the zoom to bring back the other 3 areas

You then see that it's not possible to bring more than 2 of them. In addition, while zooming in and out, areas that were out-of-zoom are moving (they are not fixed w.r.t. to the plot anymore).

I used the demo code you shared to record this video.

**Comment by helgasoft** [comment]
Yes, agree, doing it this way shows clearly the **serious bug present**.
You are zooming in and out and previous areas do not recover their position.
I did not notice it because I was zooming in **once** and then **moving the zoom window** back and forth to edit previous selection areas (works ok then)
BTW, for better visibility, use ```brushStyle: {borderColor: 'red', borderWidth: 2}```

**Comment by michaelhoarau** [comment]
Yes, and as I cannot know in advance what the users will do, it's hard to use this feature.
But redeveloping the whole brush behavior with some draggable graphics is too much work.
At this stage, this is a feature I can't offer to my user until I find a way.

Thanks for your suggestion for better visibility: I will definitely add this in our App!

Let me know if there's anyway I can help with this bug: I'm not familiar enough with eChart codebase to suggest a PR right now, but if you point me roughly in the right direction I can try and have a look or at least test any workaround you throw my way.

**Comment by helgasoft** [comment]
ok, I'll throw an idea for expansion to my code.

> in [datazoom](https://echarts.apache.org/en/api.html#events.datazoom) event handler:
> find if (start-end) difference has changed
> if (yes) { &nbsp;  // user has zoomed in or out
>  &nbsp;   dispatch [brush](https://echarts.apache.org/en/api.html#action.brush.brush) action with _areas_=allAreas  &nbsp; &nbsp;// recreate all areas from our storage
> } else {  &nbsp; // user moved zoom window
>  &nbsp;   // NOP, existing code will handle it
> }

No clue on how it will perform, but worth a try.

**Comment by michaelhoarau** [comment]
Thanks for your help, I'm going to have a look at it later this week and see if I can sort it out with dispatch action indeed.
Will let you know how it goes and how it performs.

**Comment by michaelhoarau** [comment]
Hi,
I found some time today to test your recommendation and I'm happy to say that the workaround works without any visible performance impact from the user perspective. I only use the `onBrushEnd` event:

```
function onBrushEndEvent(e, storedRanges, eChartRef) {
    let refreshAreas = false

    // If the stored areas and the new areas don't have the
    // same size, we add the last area to the storedRanges:
    if (e.areas.length > storedRanges.current.length) {
        storedRanges.current.push(e.areas[e.areas.length - 1])
    }

    // If they have the same size, then one area may have been 
    // modified. In all cases, we loop through all of them to identify 
    // a modified one. Note that the modified area is necessarily one
    // that has a non-null width as it is one visible given the current
    // data zoom:
    e.areas.forEach((area, index) => {
        const start = area.coordRange[0]
        const end = area.coordRange[1]
        const storedRangeStart = storedRanges.current[index].coordRange[0]
        const storedRangeEnd = storedRanges.current[index].coordRange[1]

        // The current area is not null and it has different coordinates: this 
        // looks like an update range, we update its record in storedRanges:
        if (start != end && (start != storedRangeStart || end != storedRangeEnd)) {
            storedRanges.current[index] = area
        }

        // Both coordinates are the same:
        else if (start == end) {
            refreshAreas = true
        }
    })

    // If a refresh is necessary, we trigger it:
    if (refreshAreas && eChartRef) {
        let areasList = []
        storedRanges.current.forEach((area) => {
            areasList.push({
                brushType: 'lineX',
                coordRange: [area.coordRange[0], area.coordRange[1]],
                xAxisIndex: 0
            })
        })

        eChartRef.current.getEchartsInstance().dispatchAction({
            type: 'brush',
            areas: areasList
        })
    }
}
```

So, this is a lot longer and I feel more brittle than what I just expected to do:

```
let currentRanges = []
e["areas"].forEach((area) => {
    currentRanges.push({
        "start": area.coordRange[0],
        "end": area.coordRange[1]
    })
})
```

But it works... Thanks a lot for your support and your ideas!
